### PR TITLE
Harden DB session cleanup against task cancellation (DEV-1861)

### PR DESCRIPTION
**Pull request by adavyas**
## Summary

Hardens the DB session cleanup path against task cancellation — **one** of the causes of the `idle in transaction` zombie-connection leak in [DEV-1861](https://linear.app/plastic-labs/issue/DEV-1861), not all of them (see *Coverage* below). A `pg_stat_activity` snapshot showed **101 backends stuck `idle in transaction`** via Supavisor (93 for 1–24h, worst 21h27m) across 34 tenant schemas, accruing ~5/hour — draining the pool, pinning the xmin horizon (vacuum starvation), and blocking DDL.

### Root cause

Cancellation (client disconnect, stream abort, request-scoped timeout, worker shutdown) is delivered at an `await`. Session cleanup lived in a `finally`:

```python
finally:
    await db.rollback()   # cancel re-delivered HERE
    await db.close()      # never runs
```

- `except Exception` never catches cancellation (`CancelledError` is a `BaseException`), so only the `finally` covered it.
- But the `finally`'s own `await`s are cancellable: a re-delivered `CancelledError` interrupts `ROLLBACK` before it reaches Postgres and skips `close()`, orphaning the connection mid-`BEGIN`.
- Supavisor (transaction mode) does **not** reset the orphaned backend on disconnect → it parks `idle in transaction` indefinitely.

**Where the re-delivery comes from.** A single `Task.cancel()` delivers `CancelledError` exactly once, so the old `finally` would have completed on its own — the bug needs cancellation re-delivered *during* cleanup. anyio provides exactly that: `CancelScope._deliver_cancellation` reschedules itself via `loop.call_soon` on **every event-loop tick** for as long as the task is inside the scope, and Starlette's `BaseHTTPMiddleware` (registered for every request in `src/main.py`) wraps each request in an anyio task group. So the storm is real and reachable on the whole API path.

This is **not** the held-session-across-external-call antipattern — the dominant path (post-Turbopuffer batch fetch) correctly opens the DB session *after* the embedding/vector call. The leak is purely in the cleanup path.

### Coverage — what this does and does not fix

| Class | Covered? |
|---|---|
| Cancelled request / task, process still alive (client disconnect, stream abort, timeout, graceful worker shutdown) | ✅ this PR |
| **Process death** — SIGKILL, OOM-kill, machine replacement/restart | ❌ **not fixable in Python** |

On process death no cleanup code runs at all: the socket to Supavisor just drops, and Supavisor in transaction mode still does not reset the server backend. Given the observed age distribution (93 backends aged 1–24h, worst 21h27m), an unknown but plausibly large share of the snapshot is this second class. The only thing that reclaims those is the server-side `idle_in_transaction_session_timeout` — see *Required alongside this PR*.

## Fix

Centralized in the three session entry points (`get_db`, `get_read_db`, `tracked_db`) — every subsystem (deriver, dreamer, reconciler, webhooks, streaming) opens sessions through these, so they all inherit the fix with no call-site changes.

- **`_run_to_completion`** — runs cleanup as a detached task and waits on it via `asyncio.wait` (which never cancels the inner task), re-raising the cancellation only after cleanup finishes. Survives a cancel *storm*. A single **monotonic** deadline bounds the whole loop so a wedged connection can't pin the task.
- **`_finalize_session`** — rollback then **always** close (inner `try/finally`); `invalidate()` the connection on a broken-mid-protocol `InterfaceError`/`OperationalError`/`DBAPIError` so a dead connection isn't returned to the pool; cleanup errors are caught+logged so they can't mask the original error.

Two guarantees stack: structural (`close()` runs even if `rollback()` is cut off) + shield (survives repeated re-delivery).

### Why not `anyio.CancelScope(shield=True)`

The ticket proposed it, but it **does not defer a native `asyncio.Task.cancel()`** — only anyio-scoped cancellation — verified empirically under both `asyncio.run()` and `anyio.run()`. Starlette (ASGI) and the deriver's uvloop both cancel natively, so the anyio shield would be inert exactly where we need it. The stdlib approach works in both runtimes and adds **no dependency**.

## Testing

`tests/test_dependencies.py` — **22 tests pass against a live Postgres** (`ruff` clean, `basedpyright` 0 errors):
- ROLLBACK+close on `CancelledError`; on cancel re-delivered during cleanup; under a cancel **storm**; broken-connection → `invalidate()`; DI-generator teardown.
- `request_context` reset even on the cancelled path (no contextvar leak in reused deriver tasks).
- Wedged-cleanup is **abandoned** past a timeout rather than pinning the task forever — including while cancellation is re-delivered at anyio's cadence.
- **Real-DB wire chaos test**: a real in-flight write transaction cancelled mid-flight leaves **no** `idle in transaction` backend (asserted via `pg_stat_activity`).

`tests/test_db_resilience.py` — 11 pass, no regression.

### Review pass (addressed)

An adversarial correctness review + a ticket-compliance audit were run. Fixes applied:
- **Bounded cleanup**: shielded cleanup is uninterruptible by design, so a wedged connection (dead socket, no libpq timeout) could pin the task forever — now abandoned past `_CLEANUP_TIMEOUT_SECONDS`.
- **Monotonic deadline** (review round 2): that timeout was passed to *each* `asyncio.wait()` call, so it re-armed on every re-delivered cancel and never elapsed — dead exactly on the anyio/`BaseHTTPMiddleware` path it was written for. Now one deadline is computed before the loop. The storm test also re-cancels at anyio's `call_soon` cadence; sparse cancels leave quiet ticks that let a per-wait timeout re-arm, which is why the old test passed either way.
- **Context reset** moved to its own `finally` so a cancelled cleanup can't leak the task-scoped contextvar.
- Guarded `fut.cancelled()`/`fut.exception()` so loop shutdown can't raise from the tail.

### Still to verify before merge

- [ ] **HTTP-level** chaos: cancel real search / chat-stream requests via client disconnect against a running server (follow-up; no harness is committed in this PR). The committed real-DB test covers the wire level; this would add the full request path.
- [ ] Prod watch: idle-in-transaction count ~0 over 24h post-deploy.

### Required alongside this PR (not code — do these regardless)

The process-death class above is only reclaimed server-side, so the first item is a companion to this PR rather than optional follow-up:

- **Server-side backstop**: `ALTER ROLE postgres SET idle_in_transaction_session_timeout = '2min';` — one line, covers every class including the ones no application fix can reach.
- Immediate remediation runbook (`pg_terminate_backend` of stuck backends).
- Monitoring alert on the existing `idle_in_txn` sampler in `lock_monitoring.sql` (count > 10 or max age > 10min).

### Follow-up (not in this PR)

- Consider TCP keepalives / `tcp_user_timeout` on the psycopg `connect_args` in `db.py`, so a black-holed socket errors out (→ `invalidate()`) instead of relying on the abandon-timeout. Shared infra; out of this ticket's scope.

Part of DEV-1861 — leaving the ticket open until the server-side timeout is applied and the 24h prod watch confirms idle-in-transaction ~0.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved cancellation-safe, bounded-time database session cleanup for request-scoped and read-only database dependencies.
  * Ensured rollback/close reliably execute even during cancellation storms and cancellation during rollback.
  * Added connection invalidation when rollback fails with database/driver rollback errors.
  * Prevented stale request-scoped context and reduced risk of lingering “idle in transaction” states after interruptions.
* **Tests**
  * Expanded database dependency tests to simulate rollback/close failures, cancellation scenarios, cleanup timeouts, and transaction-state recovery.

### CODE REVIEW

**Review by ulyssepence — COMMENTED** [review]
I read over the code in `src/dependencies.py` and it looks good to me. Interesting how `asyncio.wait` can dodge its coroutine being canceled!

The only thing that I wasn't clear on from the ticket is how confident we are that this is the problem. Was there something like a smoking gun or was this more a best guess?

**Comment by karaokedjodua** [comment]
Finn-loop review of 3bf7bd4

CI: not configured (no required checks reported on the branch)
Mergeability: BLOCKED (stale branch-protection rule — mergeable: MERGEABLE, no required checks)

## Review
Summary: Hardens all three DB session entry points (`get_db`, `get_read_db`, `tracked_db`) against task cancellation by shielding the ROLLBACK+close cleanup path via `_run_to_completion` (detached task + `asyncio.wait` loop). An added `_finalize_session` helper ensures `close()` runs even when `rollback()` is interrupted, and invalidates connections that hit mid-protocol errors. The 10s abandon timeout prevents a wedged connection from pinning the task forever. Applies uniformly across the entire codebase with zero call-site changes.

## 1. Must fix before merge

**[CI]** No required CI is configured on this branch — `gh pr checks 908 --required` reports "no required checks reported on the 'adavya/dev-1861' branch". Merge state is `BLOCKED` with `mergeable: MERGEABLE`, indicating a stale branch-protection rule referencing a check that no longer exists or matches. An admin must update or remove the stale rule before this can merge. (This is a repo-configuration gap, not a code defect.)

## 2. Should fix soon

None — the code changes are clean and well-tested.

## 3. CI check details (not required)

- **test-python**: FAIL (1m12s) — pre-existing failure in `tests/crud/test_document.py::test_duplicate_rejection_reinforces_existing` (`ValueError: OpenAI API key is required`), unrelated to this PR's changes (`src/dependencies.py`, `tests/test_dependencies.py`).
- **test-status**: FAIL (3s) — implausibly short duration; likely a CI setup issue, not a code failure.
- **CodeRabbit**: PASS
- **changes**: PASS

## 4. Code review

### Contract compliance ✓
Per the PR body (no formal Paperclip issue linked — Linear DEV-1861): the fix addresses the `idle in transaction` zombie-connection leak with the claimed approach. ACs from the PR body:
- ✅ ROLLBACK+close runs under cancellation → `_run_to_completion` shields via `asyncio.wait`
- ✅ Survives cancel re-delivery during cleanup → loop re-catches `CancelledError`
- ✅ Invalidates broken connections → `_finalize_session` catches `InterfaceError`/`OperationalError`/`DBAPIError`
- ✅ Contextvar reset survives cancel → nested `finally` in `tracked_db`
- ✅ Abandon timeout for wedged cleanup → 10s `_CLEANUP_TIMEOUT_SECONDS`

### Code quality
- No `anyio.CancelScope(shield=True)` antipattern (empirically verified ineffective against native `asyncio.Task.cancel()`)
- `_finalize_session` inner `finally` for `close()` is the correct structural guarantee
- No cross-repo fork dependency concerns — all new symbols defined within `dependencies.py`, all imports are stdlib or existing deps
- The abandoned-cleanup path logs and leaks the detached task; acceptable for a dead-socket edge case (OS/pool reclaim)

### Tests ✓
21 tests in `test_dependencies.py` (10 new for DEV-1861) plus 11 unchanged in `test_db_resilience.py`. Coverage includes: cancellation, re-delivery during rollback, cancel storm, broken-connection invalidation, contextvar reset, wedged-cleanup abandon timeout, and a real-DB `pg_stat_activity` assertion that no `idle in transaction` backend survives a cancelled write transaction.

## 5. Safe to merge

The code changes are correct, self-contained, and comprehensively tested. The merge is blocked by CI configuration (no required checks, stale branch-protection rule) — an admin needs to resolve that before merging.

---

**Labels to apply** (no admin rights on this token):
- `needs-human-review` — no required CI configured + stale branch-protection rule
- `loop-approved` — code review passes

File: src/dependencies.py [hunk 1]
```diff
@@ -1,11 +1,81 @@
...
+                "the task",
+                _CLEANUP_TIMEOUT_SECONDS,
+            )
+            break
+    if fut.cancelled():
+        cancelled = True  # cleanup itself was cancelled (close() still ran)
+    elif fut.done() and (exc := fut.exception()) is not None:
+        logger.error("session cleanup task failed", exc_info=exc)
+    if cancelled:
+        raise asyncio.CancelledError()
```
[the remark this answers is not in the document]
Comment (adavyas) [reply src/dependencies.py:61]: Fixed — the head now computes one monotonic deadline before the wait loop (`deadline = loop.time() + _CLEANUP_TIMEOUT_SECONDS`) and derives each `asyncio.wait` timeout from the remaining time, so re-delivered cancels can't re-arm it. `test_run_to_completion_deadline_holds_under_cancel_storm` asserts the hard wall-clock bound while re-cancelling at anyio's `call_soon` cadence. (The fix commit had been pushed to the upstream branch copy by mistake and never reached the PR head — now included, as of cde84654.)

File: tests/test_dependencies.py [hunk 2]
```diff
@@ -308,3 +329,238 @@ async def test_read_only_session_works_with_tracing_checkout_hook() -> None:
...
+    with pytest.raises(asyncio.CancelledError):
+        await task
+
+    state = await backend_state(pid)
+    for _ in range(20):  # allow shielded cleanup a moment to land
+        if state != "idle in transaction":
+            break
+        await asyncio.sleep(0.1)
+        state = await backend_state(pid)
+    assert state != "idle in transaction", f"backend left {state!r} after cancel"
```
[the remark this answers is not in the document]
Comment (adavyas) [reply tests/test_dependencies.py]: Fixed in cde84654 — the retry + final assertion is now wrapped in `try/finally`; the `finally` re-checks the backend from the independent observer connection and `pg_terminate_backend`s it if it's still `idle in transaction`, so a failed assertion can't park a zombie in the shared test pool.
